Integrated new dialogue-related sounds

wispykey opened the pull request:
Changes (did not split into separate commits):
1. Fixed error in HarpNPC_Interacted sequence container in WWise (there were no sounds in the playlist).
2. Play_TutorialPrompt_Activated posted when player triggers a Narrator_TriggerBox dialogue for the first time via collision.
3. Play_Dialogue_Advanced is posted when player presses the interact button for any dialogue.
4. Play_HarpNPC_Interacted is posted when player interacts with a Narrator (always a Harp so far).

Notes:
- Play_Dialogue_Advanced is currently posted on the initial interaction with a dialogue entity. Maybe disable this?
- New sounds are a bit louder than expected. This will be adjusted later.
- There is currently no distinction between narrator entities, i.e. both 'Tutorial-explaining' and 'Lore-giving' interactions play the same sound. Maybe distinguish these?
- I'm not 100% sure these are the best places to put the new AkAudio nodes. I have left the nodes in awkward positions in case they end up being moved elsewhere anyways.

BP_Narrator:
![BP_Narrator_changes](https://github.com/user-attachments/assets/37202a21-2962-40ec-af9a-f2cb620bc7eb)

BP_Narrator_Trigger_Box:
![BP_Narrator_Trigger_Box_changes](https://github.com/user-attachments/assets/455923fc-44b6-4f1b-b9fe-ca9ec8184b83)

BP_Violin (top-right corner of dialogue system):
![BP_Violin_changes](https://github.com/user-attachments/assets/6f4b7aa9-1873-4a56-a51c-16df9fe8ab57)

kseth-official commented:
Hey, I've highlighted what looks great and what needs change below!

- Fixed error in HarpNPC_Interacted sequence container in WWise (there were no sounds in the playlist). (✅) 
- Play_TutorialPrompt_Activated posted when player triggers a Narrator_TriggerBox dialogue for the first time via collision. (✅)
- Play_Dialogue_Advanced is posted when player presses the interact button for any dialogue. (❌)
- Play_HarpNPC_Interacted is posted when player interacts with a Narrator (always a Harp so far). (✅)

No Play_Dialogue_Advanced event is posted when the continue button in the DialogueSystem is pressed using the mouse, only when E is triggered. This is because triggering it in BP_Violin is incorrect.

![image](https://github.com/user-attachments/assets/30030faa-8749-47f2-8424-628bbcfdc964)

The AdvanceDialogue node in W_Dialogue is the best place to trigger Play_Dialogue_Advanced since it's the event all dialogue actors call eventually and what the continue button calls as well. I'd suggest only playing the sound when DialogueRow > 0, meaning not when the first dialogue is displayed when interaction begins (we can have a separate sound for that).

Just a quick tip for cleaning up nodes: https://dev.epicgames.com/documentation/en-us/unreal-engine/blueprint-editor-cheat-sheet-in-unreal-engine

This should help you work faster in blueprints. You can also right click->
![image](https://github.com/user-attachments/assets/cb7a4a3a-5cb0-4ee0-8536-bc15f09d7734)
to help with refactoring.

wispykey commented:
Follow up changes:

1. Play_Dialogue_Advanced is now posted in W_Dialogue instead of the BP_Violin dialogue system.
2. Cleaned up some nodes, removed most comment blocks.

Notes
- Play_Dialogue_Advanced is now correctly ignores the first interaction (by checking for dialogue row >= 1)
- Play_Dialogue_Advanced is NOT being posted on the final interaction that closes the dialogue box. Unsure if working as intended or not. 

![image](https://github.com/user-attachments/assets/f71414cd-f734-4c7f-ba67-f3f217b9f0a6)

kseth-official commented:
Awesome! Just following up with two more changes. Can we:
- use the OwningDialogueActor variable inside W_Dialogue to post the event instead of GetActorOfClass? This will post the event at the location of the DialogueActor, which here is either the harp or the trigger box, which should be okay. GetActorOfClass posts the audio to the first instance of BP_Violin found in the world, which is problematic in the case that in the future we go multiplayer and add more Violin players to the game for example.
- revert BP_Violin to the version on main (no changes) so that we don't have to resolve merge conflicts? You can git checkout the file directly from main using `git checkout main -- Content/Blueprints/Characters/Violin/BP_Violin.uasset`

wispykey commented:
Follow-up changes:

1. Ran `git checkout main -- Content/Blueprints/Characters/Violin/BP_Violin.uasset` in first commit.

2. For changing the reference in W_Dialogue, I was unable to use the OwningDialogueActor or OwningDialogueComponent variables directly (image 1). I tried casting to BFH_Char (image 2), but the casts fail (verified by inserting a print statement on output of cast fail; debug node not in image). Finally, I resorted to attempting casts to BP_Narrator and BP_Narrator_Trigger_Box separately (image 3), which works but it doesn't feel like a strictly better solution. Is there a way to avoid checking all possible types of dialogue owners? 

Variables could not be used directly:
![image](https://github.com/user-attachments/assets/c3f8db90-d7ab-4c4e-aacc-1c31d57f0c0e)


Casting either variable to BFH_Char fails.
![image](https://github.com/user-attachments/assets/3a365396-f43c-418c-8b78-f03bac606451)

Current working solution, but checks all possible owner types:
![image](https://github.com/user-attachments/assets/075b60d7-787b-4ac7-87e5-8c0c22205eb0)


3. For delay, 0.2s, 0.5s, and 20.0s were tested in BP_Narrator before ResetDialogue, but sound still did not play. The delay node was added, then removed.

kseth-official commented:
**Interesting**

Can we use the player controller directly instead? That should trigger it directly on the player instead of on the dialogue actors. I attempted it on my end, and it seems to work in AdvanceDialogue.
![image](https://github.com/user-attachments/assets/027b68f4-2b5e-4feb-955d-9ade24e71e51)

I just rechecked, and I realized that it isn't triggering when you close the dialogue box because AdvanceDialogue is called only when you're "Advancing Dialogue." Closing the dialogue box is not advancing dialogue but simply closing it. You'll need to add a separate event after the branch in BP_Narrator and BP_TriggerBox. This position indicates everything that happens when we close the dialogue box.

![image](https://github.com/user-attachments/assets/c58ca27a-7e27-4685-aa12-be64fcbff06b)

If everything works, try removing the delay nodes since it's better to clean up immediately if we can.

I'll change the type of OwningDialogue actor from DialogueInterface to just Actor, so it can be passed in directly in case we need to do anything else with the actor that owns the W_Dialogue instance.

wispykey commented:
RE: Using PlayerController0 reference

Using a reference to the controller in W_Dialogue works for dialogue in the first room of MAP_Tutorial_room, but the sound is no longer audible for any dialogue beyond the first doorway. I believe WWise requires that sounds are attached to game objects (with a position property). Posted events are only audible when the player is within a certain distance of the attached Actor - this aligns with WWise's ability to offer attenuation features. The PlayerController's phantom position is probably static near the spawn point, which would explain why sounds only play near the first room.

After basic research, it does not seem like there is an obvious way to play "object-less" sound using WWise similar to how the PlaySound2D node works.

'Play' events created in WWise must have a Scope of GameObject, i.e. be attached to a game object:
![image](https://github.com/user-attachments/assets/34612a72-1bd7-4427-8d8b-3691acfd980a)

WWise event profiler example:
![image](https://github.com/user-attachments/assets/393bbbb0-46ab-46e0-b847-0a1a23c72645)
Posting an event without an Actor input results in no event being posted at all, in the profiler.

![image](https://github.com/user-attachments/assets/e50c8932-f095-4a03-9070-7106da8e801a)


Current list of options:
1. Pass down the confirmed BP_Violin reference (named 'AsBPViolin') from BP_Narrator and BP_Narrator_Trigger_Box via the AdvanceDialogue event dispatcher. This would need to be done for every possible dialogue owner in the future, though, and assumes that the a BP_Violin's SphereCollision is what triggers the dialogue.
![image](https://github.com/user-attachments/assets/54cd5c92-e562-4c4a-9248-61cb3fd0156b)

2. Temporarily use the PlaySound2D node for dialogue-related sounds.

3. Use the first available reference to BP_Violin.

4. Attempt casts to all possible dialogue owner types.

5. Update the position of the PlayerController to match BP_Violin.

wispykey commented:
RE:  Sound on the final interaction that closes dialogue box

Would this be a reasonable solution instead of defining another event?

![image](https://github.com/user-attachments/assets/75f878c5-9a08-458c-aef4-a77ce6810c15)

wispykey commented:
Finally realized that casting OwningDialogueActor to Actor is an option.

1. Casting OwningDialogueActor to Actor (instead of attempting to cast to BP_Narrator then BP_NarratorTriggerBox).
2. Sound plays when advancing the final dialogue text. To do this, I added the DialogueFinished custom event to W_Dialogue, which is triggered in BP_Narrator and BP_NarratorTriggerBox.

kseth-official commented:
Casting OwningDialogueActor to Actor is a good solution. The dialogue finished event looks great and offers a good separation of concerns. Approved
